From 7ef8f5ebf8c7bc958e99cfd76ed84d34df078bfb Mon Sep 17 00:00:00 2001
From: Tobias Hahn <t.hahn@phytec.de>
Date: Mon, 24 Aug 2026 10:55:35 +0200
Subject: [PATCH] playbin3: Fix audio caps for GstBin as audio-sink

Use the caps of the sink pad from audio-sink instead of the caps of the
actual audio sink in the GstBin, when audio-sink is set to a GstBin,
since the GstBin might include other elements before the actual audio
sink that do not support the passthrough. This behavior also seems to
align better with the documentation and prevents bugs with Qt Multimedia
when playing a video that has audio which matches the passthrough
capabilities of the actual audio sink used by Qt Multimedia.
Keep the actual audio sink lookup logic as a fallback, since I am not
sure why the caps of the actual audio sink were used before.

Upstream-Status: Pending

Signed-off-by: Tobias Hahn <t.hahn@phytec.de>
---
 gst/playback/gstplaybin3.c | 18 +++++++++++-------
 1 file changed, 11 insertions(+), 7 deletions(-)

diff --git a/gst/playback/gstplaybin3.c b/gst/playback/gstplaybin3.c
index d16392b7e..c45910f2e 100644
--- a/gst/playback/gstplaybin3.c
+++ b/gst/playback/gstplaybin3.c
@@ -2552,7 +2552,8 @@ reconfigure_caps (GstPlayBin3 * playbin)
   }
 
   /* check the sink type */
-  if (!GST_IS_BIN ((GstBin*) playbin->audio_sink)) {
+  sinkpad = gst_element_get_static_pad (playbin->audio_sink, "sink");
+  if (!GST_IS_BIN ((GstBin*) playbin->audio_sink) || NULL != sinkpad) {
     actual_sink = playbin->audio_sink;
   } else {
     GstIterator *it;
@@ -2570,14 +2571,14 @@ reconfigure_caps (GstPlayBin3 * playbin)
     actual_sink = g_value_get_object (&item);
     g_value_unset (&item);
     gst_iterator_free (it);
-  }
 
-  if (NULL == actual_sink) {
-    GST_LOG_OBJECT (playbin, "No sink found");
-    return;
+    if (NULL == actual_sink) {
+      GST_LOG_OBJECT (playbin, "No sink found");
+      return;
+    }
+    sinkpad = gst_element_get_static_pad (actual_sink, "sink");
   }
 
-  sinkpad = gst_element_get_static_pad (actual_sink, "sink");
   if (NULL == sinkpad) {
     GST_LOG_OBJECT (playbin, "No pad found");
     return;
@@ -2586,10 +2587,13 @@ reconfigure_caps (GstPlayBin3 * playbin)
   sinkcaps = gst_pad_query_caps (sinkpad, NULL);
   if (NULL == sinkcaps) {
     GST_LOG_OBJECT (playbin, "No caps found");
+    gst_object_unref (sinkpad);
     return;
   }
 
-  GST_DEBUG_OBJECT (playbin, "sink caps: %s ", gst_caps_to_string(sinkcaps));
+  const gchar *sinkcaps_str = gst_caps_to_string(sinkcaps);
+  GST_DEBUG_OBJECT (playbin, "sink caps: %s ", sinkcaps_str);
+  g_free(sinkcaps_str);
   /* analyze the caps and update uridecodebin caps if needed */
   if (!gst_caps_is_any (sinkcaps)) {
     select_caps = gst_caps_from_string (AUDIO_PASSTHROUGH_CAPS);
-- 
2.43.0

